AE-62 Defining conventions (#2565)
spryker/spryker-docs: docs/dg/dev/architecture/coding-conventions.md

- Storefront applications (eg: Yves, Configurator, Glue, GlueStorefront) typically use the Yves-, Glue-, Client-, Service- and Shared application layers.

# Application Layers
Spryker organises responsibilities/functionalities over a set of application layers (see [Conceptual Overview](https://docs.spryker.com/docs/dg/dev/architecture/conceptual-overview.html)) to enable flexible business logic orchestration across the applications.

The application layers are aggregations of layers (see [Modules and Application Layers](https://docs.spryker.com/docs/dg/dev/architecture/modules-and-application-layers.html))
- Zed: is a **multi-layer composition** of the 4 layers: Presentation Layer, Communication Layer, Business Layer, Persistence Layer
- Yves: is a **single-layered** Presentation Layer merged with Communication Layer.
- Glue: is a **single-layered** Communication Layer.
- Client: is a **single-layered** Communication Layer.
- Service: is a **single-layered**, overarching Business Layer.
- Shared: is a layer-overarching, stateless, abstraction library.

## Zed

> Zed serves as the backend application layer responsible for housing all business logic and persisting data.

Reviewer: Shall we mention src/ and test/ directory here as well as resources/? We do have the resources folder mentioned somewhere in our confluence and by this we added it everywhere where needed.

Author: Added as follow-up: https://spryker.atlassian.net/browse/AE-94

  - Mappers MUST NOT use dependencies or restrict to [module configuration](#module-configuration), other `mappers`, or lightweight [Services](#service-facade).
  - `Persistence mappers` MUST be in [Persistence layer](#zed) only and named as `[Module|Entity]Mapper.php`.
- Retrieving and appending additional data into the existing data structure, is considered an `expander` and the method MUST be named as `expand[With<AdditionalDataExplainer>]($targetEntity)`.
- The `hydrator/hydration` keywords MUST NOT be used but instead `mapper` or `expander`.

**Guidelines**
- It is recommended to NOT define interface for `Persistence mappers`.

## Model

**Description**

Models hold logic. Models MAY appear on multiple layers, the content must fit the corresponding [layer's responsibility](#layers).
- [Communication layer](#application-layers) models are present in [Yves](#yves), [Glue](#glue), [Client](#client).
- [Business layer](#application-layers) models are present in [Zed](#zed), [Service](#service).
- [Persistence layer](#application-layers) models are present in [Zed](#zed).

**Conventions**

> - Each model class MUST define and implement an interface (`[ModelName]Interface.php`) that holds the specification of each `public` method.

Reviewer: Can't this be omitted in projects?

Author: yes, i've missed this. this is 1 of the very important project differences.

    - `Protected` constants MAY also be defined to support the getter methods (example: to enable cross-module referencing via `@uses`).
- Environment configuration
  - constants MUST be always `public` and have specification about their purpose.
  - constants MUST contain the same UPPERCASE value as the key is + properly prefixed with MODULE_NAME (see Example below).

<details><summary markdown='span'>Additional Conventions for Boilerplate, Accelerator, and Module Development</summary>
- The module configuration relays exclusively on `static::` to support extension.
</details>

<details><summary markdown='span'>Guidelines for Module Development</summary>
- `Protected` items are NOT recommended in module configuration as they tend to create BC problems on project extensions.
</details>

**Example**

```php
namespace Spryker\Shared\Oms;

interface OmsConstants
{
    /**
     * Specification:
     * - Defines if OMS transition log is enabled.
     *
     * @api
     *
     * @var string
     */
    public const ENABLE_OMS_TRANSITION_LOG = 'OMS:ENABLE_OMS_TRANSITION_LOG';
}
```

## Navigation.XML

**Description**

Represents the module entries of the Backoffice navigation panel.

> The icons are taken from [Font Awesome Icons Library](#https://fontawesome.com/v4/icon/files-o).

Reviewer: Where is the example of how to use them?

Author: i've deleted by mistake, thank you!

- The `\Spryker\[Application]\Kernel\PermissionAwareTrait` trait allows the model in the corresponding application to check if a permission is granted to an application user.

**Example**

```php
class QuotePermissionChecker implements QuotePermissionCheckerInterface
{
use PermissionAwareTrait;

    /**
     * @uses \Spryker\Client\SharedCart\Plugin\ReadSharedCartPermissionPlugin::KEY
     */
    protected const PERMISSION_PLUGIN_KEY_READ_SHARED_CART = 'ReadSharedCartPermissionPlugin';

    public function checkQuoteReadPermission(QuoteTransfer $quoteTransfer): bool
    {
        ...
      return $this->can(
        static::PERMISSION_PLUGIN_KEY_READ_SHARED_CART,
        $quoteTransfer->getCustomer()->getCompanyUserTransfer()->getIdCompanyUser(),
        $quoteTransfer->getIdQuote(),
      );
    }
```

## Persistence Schema

**Description**

The schema file defines the module's tables and columns (see [Propel Documentation - Schema XML](#https://propelorm.org/documentation/reference/schema.html)).

**Conventions**

> - Table name MUST follow the format `spy_[module]_[domain entity]` (eg: `spy_customer_address`, `spy_customer_address_book`).

Reviewer: IMO we should outline the naming format for the file as well e.g. `spy_[module].schema.xml (everything lowercase, if module name is CamelCase use and dashed e.g. product-url)`

- Tables MUST have an integer ID primary key, following the format `id_[module]_[domain entity]` (eg: `id_customer_address`, `id_customer_address_book`).

> - Table definitions MUST include the `phpName` attribute.

Reviewer: Sure? AFAIK the phpName will resolve automatically and MUST only be used when it should not be done atomatically.

Author: @stereomon May i ask you to verify if table & foreign key's name is default camelcase generated? if yes, i will just remove this line, it's deprecated then.

- Table foreign key MUST follow the format `fk_[remote entity]` (eg: `fk_customer_address`, `fk_customer_address_book`).

> - Table foreign key definitions MUST include the `phpName` attribute.

Reviewer: Same here. I see lots of places where it is missing.

Reviewer: Since we are not following these two rules, we should remove them from this doc.

Author: That is not how we need to consdier the conventions. Conventions are targeting the future. We want to make sure that future contrubtions are following a specific list, that are aligned with our long-term goals. Conventions are not present for past. I do agree that it is the desired state that our past & present & future codes are fully unified over conventions, but it is not always possible.

> - Tables with sensitive primary key (eg: `spy_order` ID gives information about submitted order count) MUST include UUID field for public access purpose.

Reviewer: Shouldn't we simply say "Identifiers used for external communication MUST use a UUID"?

```xml
<?xml version="1.0"?>
<database xmlns="spryker:schema-01" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="zed" xsi:schemaLocation="spryker:schema-01 https://static.spryker.com/schema-01.xsd" namespace="Orm\Zed\Url\Persistence" package="src.Orm.Zed.Url.Persistence">

    <table name="spy_url">
        <column name="fk_resource_product_abstract" type="INTEGER"/>
        <foreign-key name="spy_url-fk_resource_product_abstract" foreignTable="spy_product_abstract" phpName="SpyProduct" onDelete="CASCADE">
            <reference foreign="id_product_abstract" local="fk_resource_product_abstract"/>
        </foreign-key>
    </table>
</database>
```

## Provider / Router

**Description**

Providers are used during the bootstrap of [Yves](#yves). There are three types of providers:
- **Controller Provider** - Registers [Controllers](#controller) of a module and binds them to a path.
- **Service Provider** - Represents a cross-functional service (eg: a validator).
- **Router** - Resolves a path into a module, [Controller and Action](#controller).

**Conventions**

> - All providers and routers MUST be registered in `[Module]/Yves/ShopApplication/YvesBootstrap.php` (see [Silex Documentation about Providers](#https://symfony.com/blog/the-end-of-silex/doc/1.3/providers.html))

Reviewer: Shouldn't we remove Silex mentionings completely from this doc?

Yves provides only 1-level Theme inheritance: Current Theme > Default Theme.
- Current Theme - a single theme defined on a project level. Eg b2b-theme, b2c-theme.
- Default Theme - a theme provided from Spryker by default and used in the Master Suite. Used for incremental project updates (start from default and change components one-by-one) and a graceful fallback in case Spryker provides a new functionality that does not have own frontend in a project.

**Contract**
- There is always a "default" theme.
- A Theme may contain Views, Templates or Components (atoms, molecules or organisms)

## Transfer Object

**Description**

Transfer objects are pure data containers with getters and setters.
They can be used in all Applications and all Layers.
They are described in XML files and then become autogenerated into the `src/Generated/Shared/Transfer` directory.
For every defined table in [Peristence Schema](#persistence-schema) a matching `EntityTransfer` will be generated with `EntityTransfer` suffix.
`EntityTransfers` are the lightweight data transfer objects of the underlying data, and this object should be used as source of communication.

**Conventions**

> - Naming: `[module].transfer.xml`.

Reviewer: Like this also for the schema XML, worth mentioning the dashing...

**Conventions**
- The components MUST be placed according to the corresponding application layer’s directory architecture (see [Application Layers](#application-layers)) in order to come in effect.
- The components MUST be extended __only__ from the application-layer corresponding Abstract class in `Kernel` module.
- The components MUST be stateless to be deterministic and easy to comprehend.

## Bridge

**Description**

According to the Interface Segregation Principle each module defines an interface for each foreign class it depends on. Bridges are used to wrap and decouple dependencies (such as Facade, Client, Service, external library class, etc).This way it declares its own requirements and any class which implements the interface can be used. The problem appears when a facade from another module wants to implement the interface. In this case there is a cross-module coupling between that facade and the implemented interface. To avoid this situation we are using the bridge pattern. This allows easy module decoupling and both can evolve independently.

**Conventions**

- Bridges and their interfaces are defined in Dependency\{facade type}\ directory.

>   - Class name follows `[CurrentModule]To[FacadeOnwerModule]FacadeBridge`.

Reviewer: ```suggestion
  - Class name follows `[CurrentModule]To[FacadeOwnerModule]FacadeBridge`.
```

Typo

- Plugin method names MUST use words `pre` and `post` instead of `before`, `after`.
- Plugin class names MUST use words `pre`, `post`, `create`, `update`, `delete`, instead of `creator`, `updater`, `deleter`.
- The plugin MUST implement a [Plugin Interface](#plugin-interface) which is provided by an extension module.

**Guidelines**
- Plugin name should be unique and give and overview about the behaviour that the plugin delivers (consider developers searching a plugin in a catalog across the framework only by plugin name).

**Example**

```php
  /**
    * @see /F/Q/N/FooDependencyProvider::getFooPlugins()
    * @see /F/Q/N/BarDependencyProvider::getBarPlugins()
    */
  class ZipZapPlugin extends AbstractPlugin implements FooPluginInterface, BarPluginInterface {}
```

## Plugin Interface

**Description**

- Modules which consume plugins need to define their requirements with an interface.
- The plugin interface needs to be placed in an “extension module”.

There are three modules involved:

1. **Plugin definer** (aka extension module): The module that defines and holds the plugin interface (example: CompanyPostCreatePluginInterface in CompanyExtension module.

> 2. **Plugin executor**: The module that demands the plugin(s) in its Dependency Provider (example: CompanyDependencyProvider::getCompanyPostCreatePlugins() in Company module)

Reviewer: ```suggestion
2. **Plugin executor**: The module that uses the plugin(s) in its Dependency Provider (example: CompanyDependencyProvider::getCompanyPostCreatePlugins() in Company module)
```
Demands sounds wrong here to me.

3. **Plugin providers**: The modules that implement a plugin (example: CompanyBusinessUnitCreatePlugin in CompanyBusinessUnit module)

**Conventions**
- Plugin interfaces MUST be defined in extension module.
  - Extension modules MUST be suffixed with Extension and follow regular module architecture.
  - Extension modules MUST NOT contain anything else but plugin interfaces and Shared layer.
- Plugin interface MUST have a class specification with plugin-stack details.
- Each method of a plugin receives a collection as an input argument.

**Guidelines**

> - Operations on single items in plugin stack is forbidden except for the following reasons (check this RFC for more details):

Reviewer: Don't forget to add the link here.

Author: I've removed instead. RFC should have been integrated already.

- Boilerplate and Accelerator Contribution: Contributing to boilerplates or accelerators requires additional considerations. We provide guidelines specific to this use-case to streamline your contributions.
- Module Contribution: For those contributing to modules, there are specialized rules to follow within the module folders. This ensures consistency and compatibility across SCOS.

## Directive Classification
Throughout this documentation, you will encounter two types of directives:
- Convention: These are mandatory requirements that contributors must adhere to in order to enable specific SCOS features or ensure proper application responses.
- Guideline: While not mandatory, following these guidelines is highly recommended. Doing so promotes long-term code maintainability and facilitates smoother development processes.

# Applications
Spryker utilises application layers into applications to enable constructing the necessary application architecture for the specific business requirements to provide a quick project start, and long-term maintainability.

> - Backend applications (eg: Zed, Backend API, Backend Gateway, Backoffice, GlueBackend, MerchantPortal) typically use the Zed-, Glue-, Client-, Service-, and Shared application layers.

Author: @pyvovarov-s :) that's funny. What you see above, is the actual list of our existing apps in public folder :)

> - Backend applications (eg: Zed, Backend API, Backend Gateway, Backoffice, GlueBackend, MerchantPortal) typically use the Zed-, Glue-, Client-, Service-, and Shared application layers.

Reviewer 2: I'm not sure why we have 2 of those, but checked and you are right, this is an exact public folder

## Presentation Layer responsibilities
- Responsible for handling the user interface (UI) presentation.
- Contains Twig templates, JavaScript (JS), and Cascading Style Sheets (CSS) files.
- Handles user interactions and input validations on the client side.
- Interacts with the [Communication Layer](#communication-layer-responsibilities) to retrieve necessary data for display.

## Communication Layer responsibilities
- Acts as an intermediary between the [Presentation Layer](#presentation-layer-responsibilities) and the [Business Layer](#business-layer-responsibilities).
- Contains [controllers](#controller) responsible for handling HTTP requests and responses.
- Contains [plugins](#plugin) responsible for flexible, overarching requests and responses.
- Manages form processing and validation.
- Handles [routing and dispatching requests](#provider--router) to appropriate [controllers](#controller).
- Interacts with the [Business Layer](#business-layer-responsibilities) to perform business operations.

## Business Layer responsibilities
- Contains the main business logic.
- Implements business rules and processes.
- Performs data manipulation, calculations, and validation.

> - Interacts with the [Persistence Layer](#persistence-layer-responsibilities) to read and write data.

Author: we don't have such thing in our conventions. do you have maybe a source that can be processed quickly?

- Independent of the user interface and communication protocols.

## Persistence Layer responsibilities
- Responsible for data storage and retrieval.

> - Contains queries (via [Entity Manager](#entity-manager) or [Repository](#repository)), [entities](#entity) (data models), and [database schema definitions](#persistence-schema).

Author: mappers are present everywhere. they are generic design patterns, not part of our infrastructure

> {% info_block infoBox "Info" %}

Author: Rename the page to Architectural Conventions to better fit the content

This document serves as a dynamic resource intended to provide an overview of our conventions and guidelines. We acknowledge that the technology landscape is ever-evolving, and as such, this document is subject to continuous refinement and improvement.

Your feedback and suggestions are highly valued to enhance the accuracy, relevance, and effectiveness of Spryker Commerce Operating System (SCOS). We encourage you to contribute your insights and recommendations by proposing changes through our designated channels.

{% endinfo_block %}

# Documentation structure

> ## Development Use-Cases

Author: Add more clarity on the convention structure to the readers to make it simpler to understand what conventions apply for a given case.

{% info_block infoBox "Info" %}

This document serves as a dynamic resource intended to provide an overview of our conventions and guidelines. We acknowledge that the technology landscape is ever-evolving, and as such, this document is subject to continuous refinement and improvement.

Your feedback and suggestions are highly valued to enhance the accuracy, relevance, and effectiveness of Spryker Commerce Operating System (SCOS). We encourage you to contribute your insights and recommendations by proposing changes through our designated channels.

{% endinfo_block %}

# Documentation structure

## Development Use-Cases
Understanding the development scenarios in which Spryker Commerce Operating System (SCOS) can be utilized is crucial for maximizing its potential. We have outlined specific behaviors and guidelines tailored to different use-cases:

- **Project Development**: If you are developing a project, you will need to adhere to specific project development guidelines to ensure smooth integration.

> - **Boilerplate and Accelerator Contribution**: Contributing to boilerplates or accelerators requires additional considerations. We provide guidelines specific to this use-case to streamline your contributions.

Reviewer 3: ```suggestion
- **Module Development**: Contributing 3rd-party reusable modules, boilerplates or accelerators requires additional considerations. We provide guidelines specific to this use-case to streamline your contributions. These guidelines are stricter than Project Development, as such functionalities should be re-usable on multiple projects in different contexts.
```

> # Table of Content

Reviewer 4: This can be removed as we have "Table of Content" as it is a part of documentation website functionality

- [Module Configurations](#module-configurations)
- [Navigation.XML](#navigationxml)
- [Permission Plugin](#permission-plugin)
- [Persistence Schema](#persistence-schema)
- [Provider / Router](#provider--router)
- [Query Container facade](#query-container-facade)
- [Query Object](#query-object)
- [Repository](#repository)
- [Service facade](#service-facade)
- [Theme](#theme)
- [Transfer Object](#transfer-object)
- [Widget](#widget)
- [Zed Stub](#zed-stub)
- [Bridge](#bridge)
- [Plugin](#plugin)
- [Plugin Interface](#plugin-interface)

# Documentation structure

## Development Use-Cases
Understanding the development scenarios in which Spryker Commerce Operating System (SCOS) can be utilized is crucial for maximizing its potential. We have outlined specific behaviors and guidelines tailored to different use-cases:

- **Project Development**: If you are developing a project, you will need to adhere to specific project development guidelines to ensure smooth integration.
- **Boilerplate and Accelerator Contribution**: Contributing to boilerplates or accelerators requires additional considerations. We provide guidelines specific to this use-case to streamline your contributions.

> - **Module Contribution**: For those contributing to modules, there are specialized rules to follow within the module folders. This ensures consistency and compatibility across SCOS.

Reviewer 5: mb we can specify that we are talking about Spryker's modules?

## Directive Classification
Throughout this documentation, you will encounter two types of directives:
- **Convention**: These are mandatory requirements that contributors must adhere to in order to enable specific SCOS features or ensure proper application responses.
- **Guideline**: While not mandatory, following these guidelines is highly recommended. Doing so promotes long-term code maintainability and facilitates smoother development processes.

# Applications

> Spryker utilises application layers into applications to enable constructing the necessary application architecture for the specific business requirements to provide a quick project start, and long-term maintainability.

Reviewer 4: We do not have a console described here

> - Backend applications (eg: Zed, Backend API, Backend Gateway, Backoffice, GlueBackend, MerchantPortal) typically use the Zed-, Glue-, Client-, Service-, and Shared application layers.

Reviewer 5: I read your conversation, but I would still name application, not by the folder name, but by their name. Also mb we can put references to pages that are describing those applications or at least add additional context. E.g. new Glue Infrastructure page or decoupled application for Zed

Author: could you please propose it in a diff?

- Storefront applications (eg: Yves, Configurator, Glue, GlueStorefront) typically use the Yves-, Glue-, Client-, Service- and Shared application layers.

# Application Layers
Spryker organises responsibilities/functionalities over a set of application layers (see [Conceptual Overview](https://docs.spryker.com/docs/dg/dev/architecture/conceptual-overview.html)) to enable flexible business logic orchestration across the applications.

The application layers are aggregations of layers (see [Modules and Application Layers](https://docs.spryker.com/docs/dg/dev/architecture/modules-and-application-layers.html))
- **Zed**: is a **multi-layer composition** of the 4 layers: `Presentation Layer`, `Communication Layer`, `Business Layer`, `Persistence Layer`.

> - **Yves**: is a **single-layered** `Presentation Layer` merged with `Communication Layer`.

Reviewer 6: I don't understand why we consider Yves as single-layer?

Author: The layers are not distingiushed from each other. You can't find the `Presenttation` or `Communication` directories, like in Zed. I will use another term, i see why the `single-layered` a bit confusing.

- **Glue**: is a **single-layered** `Communication Layer`.
- **Client**: is a **single-layered** `Communication Layer`.
- **Service**: is a **single-layered**, overarching `Business Layer`.
- **Shared**: is a layer-overarching, stateless, abstraction library.

## Zed

> `Zed` serves as the backend application layer responsible for housing all business logic and persisting data.

Reviewer 7: @Vetaxon what is `conventional APIs of `Shared` layer`?

> `Zed` serves as the backend application layer responsible for housing all business logic and persisting data.

Reviewer 7: Zed can also call 3rd party libraries, another Zed modules.
It cannot make a Gateway RPC call.

> `Zed` serves as the backend application layer responsible for housing all business logic and persisting data.

Reviewer 8: Each layer has its own interface for communication 
Client - Client
Service - Service 
Shared - Config, class const refs ..

```
[Organization]
└── Zed
    └── [Module]
        ├── Presentation
        │   ├── Layout
        │   │   └── [layout-name].twig
        │   └── [name-of-controller]
        │       └── [name-of-action].twig
        │    
        ├── Communication
        │   ├── Controller
        │   │   ├── IndexController.php
        │   │   ├── GatewayController.php
        │   │   └── [Name]Controller.php
        │   ├── [CustomDirectory]
        │   │   ├── [ModelName]Interface.php
        │   │   └── [ModelName].php        
        │   ├── Exception
        │   ├── Plugin
        │   │   └── [ConsumerModule]
        │   │       └── [PluginName]Plugin.php
        │   ├── Form
        │   │   ├── [Name]Form.php
        │   │   └── DataProvider
        │   │       └── [Name]FormDataProvider.php
        │   ├── Table
        │   │   └── [Name]Table.php                
        │   ├── navigation.xml       

>         │   └── [Module]CommunicationFactory.php

Reviewer 9: these folders do not fit the description above 

>`Zed` serves as the backend application layer responsible for housing all business logic and persisting data.

So update the description and include UI for Backoffice.

        │            
        ├── Business
        │   ├── [CustomDirectory]
        │   │   ├── [ModelName]Interface.php
        │   │   └── [ModelName].php        
        │   ├── Exception

>         │   ├── Installer

Reviewer 5: do we really have this? what a purpose of `Installer` directory?

        │   │   ├── [Module]InstallerInterface.php
        │   │   └── [Module]Installer.php        
        │   ├── [Module]BusinessFactory.php
        │   ├── [Module]Facade.php
        │   └── [Module]FacadeInterface.php
        │
        ├── Persistence
        │   ├── [Module]EntityManagerInterface.php
        │   ├── [Module]EntityManager.php
        │   ├── [Module]QueryContainer.php
        │   ├── [Module]PersistenceFactory.php
        │   ├── [Module]RepositoryInterface.php
        │   ├── [Module]Repository.php
        │   └── Propel
        │       ├── Abstract[TableName].php
        │       ├── AbstractSpy[TableName]Query.php
        │       ├── Mapper
        │       │   └── [Module|Entity]Mapper.php
        │       └── Schema
        │           └── [organization_name]_[domain_name].schema.xml
        │

>         ├── Dependency

Reviewer 5: in theory QueryContainers also can be here
Also [Module]Events.php can be seen here in some modules

Author: you are not wrong, this needs to be added to our conventions separately. i can't do it during this transfer, unless you could provide the description for Events.php

        │   ├── Client

>         │   │   ├── [Module]To[Module]ClientBridge.php

Reviewer 5: Bridge|Adapter

Author: although you are 100% right here, i don't want to advertise the Adapters, as they are rare cases. i want to avoid that people start to define Adapters by mistake.

        │   │   └── [Module]To[Module]ClientInterface.php          
        │   ├── Facade
        │   │   ├── [Module]To[Module]FacadeBridge.php
        │   │   └── [Module]To[Module]FacadeInterface.php
        │   ├── Service
        │   │   ├── [Module]To[Module]ServiceBridge.php
        │   │   └── [Module]To[Module]ServiceInterface.php      

>         │   └── Plugin

Reviewer 5: IMO this is mostly exception. I can remember only MoneyPlugin as an example of such dependency

Author: We have 69 plugin stack definition in Yves (eg: `SprykerShop\Yves\SessionCustomerValidationPageExtension\Dependency\Plugin\CustomerSessionValidatorPluginInterface`). Or maybe i did not understand your point?

- [Gateway Controller](#gateway-controller)
- [Layout](#layout)
- [Mapper / Expander / Hydrator](#mapper--expander--hydrator)
- [Models](#model)
- [navigation.xml](#navigationxml)
- [Plugin](#plugin), [Plugin Interface](#plugin-interface)
- [Query Container](#query-container-facade)
- [Query Object](#query-object)
- [Repository](#repository)
- [Schema](#persistence-schema)

## Yves
`Yves` application layer provides a lightweight Shop Application.

```
[Organization]
└── Yves
    └── [Module]
        ├── Controller
        │   ├── IndexController.php
        │   └── [Name]Controller.php
        ├── [CustomDirectory]
        │   ├── [ModelName]Interface.php
        │   └── [ModelName].php                
        ├── Dependency
        │   ├── Client
        │   │   ├── [Module]To[Module]ClientBridge.php
        │   │   └── [Module]To[Module]ClientInterface.php
        │   ├── Service
        │   │   ├── [Module]To[Module]ServiceBridge.php
        │   │   └── [Module]To[Module]ServiceInterface.php           
        │   └── Plugin
        │       └── [PluginInterfaceName]PluginInterface.php
        ├── Plugin
        │   ├── Provider
        │   │   ├── [Name]ControllerProvider.php

>         │   │   └── [Name]ServiceProvider.php

Reviewer 5: what is this ServiceProvider controller?

Author: it's not a controller. as it is described a few lines below:
- **Service Provider** - Represents a cross-functional [Yves](#applications) service (eg: a validator).
- A `Service Provider` MUST implement `\Silex\ServiceProviderInterface and may extend \Spryker\Yves\Kernel\AbstractPlugin`.

Based on this, you can find 
- TwigMoneyServiceProvider
- TranslationServiceProvider

        │       │   │   └── [atom-name]
        │       │   │       └── [atom-name].twig
        │       │   └── molecules
        │       │       └── [molecule-name]
        │       │           └── [molecule-name].twig
        │       ├── templates
        │       │   ├── page-layout-[page-layout-name]
        │       │   │   └── page-layout-[page-layout-name].twig
        │       │   └── [template-name]
        │       │       └── [template-name].twig
        │       └── views
        │           └── [name-of-controller]
        │               └── [name-of-action].twig        
        ├── Widget
        │   └── [Name]Widget.php
        ├── [Module]Config.php
        ├── [Module]DependencyProvider.php
        └── [Module]Factory.php
```

Used components
- [Bridge](#bridge)
- [Config](#module-configurations)
- [Controller](#controller)
- [Dependency Provider](#dependency-provider)
- [Factory](#factory)
- [Layout](#layout) 
- [Mapper / Expander / Hydrator](#mapper--expander--hydrator)
- [Model](#model)
- [Provider / Router](#provider--router)
- [Plugin](#plugin), [Plugin Interface](#plugin-interface)
- [Templates](#theme)
- [Theme](#theme)
- [Widget](#widget)

## Glue

> The `Glue` application layer serves as a conduit for providing data access points through APIs.

Reviewer 7: @Vetaxon which is not true for Backend Glue API.

> The `Glue` application layer serves as a conduit for providing data access points through APIs.

Reviewer 8: edited

> The `Glue` application layer serves as a conduit for providing data access points through APIs.

Author: I don't want to mix in our custom applications. Although it's crucial to understand that GlueBackend supposed to call only Zed, the current page wants to talk about the framework itself; where you can put together your applications, layers, etc. as you wish. But it's good thinking 👍

It acts as an interface for external systems to interact with the application's data.

```
[Organization]
└── Glue
    └── [Module]
        ├── Controller
        │   ├── IndexController.php        
        │   └── [Name]Controller.php                
        ├── [CustomDirectory]
        │   ├── [ModelName]Interface.php
        │   └── [ModelName].php        

>         ├── Dependency

Reviewer 5: Facade also can present in Glue module if it's BackendApi module

Author: excellent point!

        │   ├── Client
        │   │   ├── [Module]To[Module]ClientBridge.php
        │   │   └── [Module]To[Module]ClientInterface.php
        │   ├── Service
        │   │   ├── [Module]To[Module]ServiceBridge.php
        │   │   └── [Module]To[Module]ServiceInterface.php           
        │   └── Plugin
        │       └── [PluginInterfaceName]PluginInterface.php
        ├── Plugin
        │   ├── Provider
        │   │   ├── [Name]ControllerProvider.php
        │   │   └── [Name]ServiceProvider.php
        │   └── [ConsumerModule]
        │       └── [PluginName]Plugin.php
        ├── [Module]Config.php
        ├── [Module]DependencyProvider.php
        └── [Module]Factory.php
```

Used components
- [Bridge](#bridge)
- [Config](#module-configurations)
- [Controller](#controller)
- [Dependency Provider](#dependency-provider)
- [Factory](#factory)
- [Mapper / Expander / Hydrator](#mapper--expander--hydrator)
- [Model](#model)
- [Plugin](#plugin), [Plugin Interface](#plugin-interface)

## Client

`Client` is a lightweight application layer that handles all data access, such as
- Persistence access (such as Key-Value Storage (Redis), Search (Elasticsearch)),

> - Yves sessions,

Reviewer 4: Session is a business logic use case of Persistence access

- `Zed` as a data-source (RPC),

> - 3rd party communication.

Reviewer 7: Zed was already mentioned 2 line above.

> - 3rd party communication.

Reviewer 8: edited

Note: Backend database access is an exception for performance streamlining.

```
[Organization]
└── Client
    └── [Module]
        ├── [CustomDirectory]
        │   ├── [ModelName]Interface.php
        │   └── [ModelName].php        
        ├── Dependency
        │   ├── Client
        │   │   ├── [Module]To[Module]ClientBridge.php
        │   │   └── [Module]To[Module]ClientInterface.php
        │   ├── Service
        │   │   ├── [Module]To[Module]ServiceBridge.php
        │   │   └── [Module]To[Module]ServiceInterface.php           
        │   └── Plugin
        │       └── [PluginInterfaceName]PluginInterface.php            
        ├── Plugin
        │   └── [ConsumerModule]
        │       └── [PluginName]Plugin.php
        ├── Zed
        │   ├── [Module]Stub.php
        │   └── [Module]StubInterface.php

>         ├── [CustomDirectory]

Reviewer 5: I guess we don't need this line

        ├── [Module]ClientInterface.php
        ├── [Module]Client.php
        ├── [Module]Config.php
        ├── [Module]DependencyProvider.php
        └── [Module]Factory.php
```

Used components
- [Bridge](#bridge)
- [Client facade](#client-facade)
- [Config](#module-configurations)
- [Dependency Provider](#dependency-provider)
- [Factory](#factory)
- [Mapper / Expander / Hydrator](#mapper--expander--hydrator)
- [Model](#model)
- [Plugin](#plugin), [Plugin Interface](#plugin-interface)
- [Zed Stub](#zed-stub)

## Service

The `Service` application layer encapsulates a multipurpose library that can be utilized across various application layers, such as `Yves` or `Zed` application layers.
A `Service` primarily consists of reusable lightweight stateless business logic components.

> Due to its deployment across all applications, a `Service` is constrained to accessing data providers that are available universally (for instance, the backend database is not accessible from storefront applications by default).

Author: we do, but resources are not wired to all applications.

If you write a `DBService::saveDBData($table, $row)`, then you can call this service from Yves & Zed, because the Service code is deployed in all applications, but Yves will throw you an exception when it can't reach the relational database.

```
[Organization]
└── Service
    └── [Module]
        ├── [CustomDirectory]
        │   ├── [ModelName]Interface.php
        │   └── [ModelName].php        
        ├── [Module]Config.php            
        ├── [Module]DependencyProvider.php
        ├── [Module]ServiceFactory.php
        ├── [Module]ServiceInterface.php
        └── [Module]Service.php
```

Used components
- [Config](#module-configurations)
- [Dependency Provider](#dependency-provider)
- [Factory](#factory)
- [Mapper / Expander / Hydrator](#mapper--expander--hydrator)
- [Model](#model)
- [Service facade](#service-facade)

## Shared

The `Shared` library contains code that is designed to be utilized across different application layers and modules.

> The `Shared` library is intended to facilitate the sharing of code among various application layers and modules.

Author: Service shares usable services. Shared, shares code. Eg: if you have an abstract base class, that must be the same across all application layers, but it's implementation depends on the application layer, then you need to save the abstract class in Shared. Service can not be wired differently per application layer.

> To ensure compatibility and versatility across different application architecture setups, any content within the `Shared` library must be free of application layer-specific elements. Therefore, the use of [Factories](#factory) is not permitted within the `Shared` library.

Reviewer 5: mb we can also mention that shared config can be accessed from Application Layer specific configs

```
[Organization]
└── Shared
    └── [Module]
        ├── Transfer
        │   └── [module_name].transfer.xml
        ├── [Module]Constants.php
        └── [Module]Config.php
```

Used components
- [Config](#module-configurations)
- [Constants](#module-configurations)
- [Transfer](#transfer-object)

# Layers

An application layer can have up to four logical layers with clear purpose and communication rules.

> - **Presentation Layer**: contains twig-templates, JS and CSS files

Reviewer 2: Same as previous, it's not limited to that we have ts, less ...

> - **Communication Layer**: contains controllers, forms, tables and plugins

Reviewer 2: Also is not limited to this types, data providers ...

Author: so how would you phrase this line?

> - **Communication Layer**: contains controllers, forms, tables and plugins

Author: I've added the `console commands`. I can't add `etc.` because it gives no limitation what can be included.

- **Business Layer**: contains the business logic of a module
- **Persistence Layer**: contains queries, entities and the schema of a module

## Presentation Layer responsibilities
- Responsible for handling the user interface (UI) presentation.
- Contains frontend elements, such as HTML, Twig templates, JavaScript (JS), TypeScript, Cascading Style Sheets (CSS) files, and so on.
- Handles user interactions and input validations on the client side.
- Interacts with the [Communication Layer](#communication-layer-responsibilities) to retrieve necessary data for display.

## Communication Layer responsibilities
- Acts as an intermediary between the [Presentation Layer](#presentation-layer-responsibilities) and the [Business Layer](#business-layer-responsibilities).

> - Contains [controllers](#controller) responsible for handling HTTP requests and responses.

Reviewer 6: ```suggestion
- Contains console commands.
- Contains [controllers](#controller) responsible for handling HTTP requests and responses.
```

Reviewer 8: It is mentioned bellow about plugins 
Console command is a plugin as well

Reviewer 6: I have been working in Spryker for three years, and it is not clear for me that Console command is a plugin. We should explicitly mention that either here, or under the plugins.

- Contains [plugins](#plugin) responsible for flexible, overarching requests and responses.
- Manages form processing and validation.
- Handles [routing and dispatching requests](#provider--router) to appropriate [controllers](#controller).
- Interacts with the [Business Layer](#business-layer-responsibilities) to perform business operations.

## Business Layer responsibilities
- Contains the main business logic.
- Implements business rules and processes.
- Performs data manipulation, calculations, and validation.
- Interacts with the [Persistence Layer](#persistence-layer-responsibilities) to read and write data.
- Independent of the user interface and communication protocols.

## Persistence Layer responsibilities
- Responsible for data storage and retrieval.
- Contains queries (via [Entity Manager](#entity-manager) or [Repository](#repository)), [entities](#entity) (data models), and [database schema definitions](#persistence-schema).
- Handles database operations such as CRUD (`Create`, `Read`, `Update`, `Delete`).
- Ensures data integrity and security.

# Components

**Conventions**

> - The components MUST be placed according to the corresponding [application layer’s](#application-layers) directory architecture  in order to come in effect.

Reviewer 2: Sometimes MUST, and should capitalized and sometimes they are not, I think we need to align them

- The components MUST inherit from the [application layer](#application-layers) corresponding abstract class in `Kernel` module.
- The components MUST be stateless to be deterministic and easy to comprehend.

<details><summary markdown='span'>Additional Conventions for Module Development</summary>
- The components MUST be extended directly from the [application layer](#application-layers) corresponding abstract class in `Kernel` module.<br/>
</details>

> ## Client facade

Reviewer 2: In some places of the doc, it's just Client and in some, it's Client facade, maybe we should align  that

> ## Client facade

Reviewer 4: Can we just merge all facade use cases into one description like controllers and dependency providers?

```
[Organization]
└── Client
    └── [Module]
        ├── [Module]ClientInterface.php
        └── [Module]Client.php         
```

**Description**

> The `Client` is the internal API of [Client application layer](#client), and acts as the only entry point to the [Client Application](#applications).

Reviewer 2: I'm not sure that `internal` word is described somewhere, it looks like we can omit this or explain

     * items were added in total to a specific slot.
     * - Returns QuoteResponseTransfer.
     *
     * @api
     *
     * @param \Generated\Shared\Transfer\CreateConfiguredBundleRequestTransfer $createConfiguredBundleRequestTransfer
     *
     * @return \Generated\Shared\Transfer\QuoteResponseTransfer
     */
    public function addConfiguredBundle(CreateConfiguredBundleRequestTransfer $createConfiguredBundleRequestTransfer): QuoteResponseTransfer;
}
```

## Controller

```
[Organization]
├── Zed
│   └── [Module]
│       └── Communication
│           └── Controller
│               ├── IndexController.php
│               ├── GatewayController.php
│               └── [Name]Controller.php
│
├── Yves
│   └── [Module]
│       └── Controller
│           ├── IndexController.php
│           └── [Name]Controller.php    
│
└── Glue
    └── [Module]
        └── Controller
            ├── IndexController.php        
            └── [Name]Controller.php                      
```
**Description**

`Controllers` and `Actions` are application access points for any kind of HTTP communication with end-users or other applications.

Responsibilities of a controller are

> - to adapt the received input data to the underlying layers (syntactical validation, delegation),

Reviewer 6: Do we do any kind of validation in the controller?

```suggestion
- to pass the received input data to the underlying layers,
```

Author: We should :) 
https://en.wikipedia.org/wiki/Model%E2%80%93view%E2%80%93controller

- delegate the processing of the input data,
- and to adapt the results of processing to the expected output format (eg: add flash messages, set response format, trigger redirect).

The `Gateway` controller name is reserved for the [Gateway Controller](#gateway-controller).

The `Index` controller name acts as the default controller during request controller resolution.

The `index` action name acts as the default action during request action resolution.

**Conventions**
- `Public` methods are considered as an `Action` and therefore MUST use the `Action` suffix.
- Action methods MUST have either no parameter or receive the `\Symfony\Component\HttpFoundation\Request` object to access system or request variables.

> - Received data MUST be [syntactically validated](#https://cheatsheetseries.owasp.org/cheatsheets/Input_Validation_Cheat_Sheet.html).

Reviewer 6: What is the scope of this validation?

Reviewer 8: I am confused with the message as well

Author: I have improved the message.

class TemplateController extends Spryker\Zed\ConfigurableBundleGui\Communication\Controller\TemplateController
{
    /**
     * @var string
     */
    protected const PARAM_ID_CONFIGURABLE_BUNDLE_TEMPLATE = 'id-configurable-bundle-template';
    
    /**
     * @param \Symfony\Component\HttpFoundation\Request $request
     *
     * @return \Symfony\Component\HttpFoundation\JsonResponse
     */
    public function slotTableAction(Request $request): JsonResponse
    {
        $idConfigurableBundleTemplate = $this->castId(
            $request->get(static::PARAM_ID_CONFIGURABLE_BUNDLE_TEMPLATE),
        );

        $table = $this->getFactory()->createConfigurableBundleTemplateSlotTable($idConfigurableBundleTemplate);

        return $this->jsonResponse($table->fetchData());
    }
}
```

## Dependency Provider

```
[Organization]
├── Zed
│   └── [Module]        
│       ├── [Module]Config.php
│       └── [Module]DependencyProvider.php
├── Yves
│   └── [Module]        
│       ├── [Module]Config.php
│       └── [Module]DependencyProvider.php
├── Glue
│   └── [Module]        
│       ├── [Module]Config.php
│       └── [Module]DependencyProvider.php
├── Client
│   └── [Module]        

> │       ├── [Module]Config.php

Reviewer 5: config seems a bit unnecessary here

│       └── [Module]DependencyProvider.php
└── Service
    └── [Module]        
        ├── [Module]Config.php
        └── [Module]DependencyProvider.php
```

**Description**

Injects required dependencies to a module application layer. Typically, dependencies are [facades](#facade-design-pattern) or [plugins](#plugin). 

Dependency injection is orchestrated through a `provide-add-get` structure (see example below).
- The `provide` method is the highest level that holds only `add` calls without any additional logic.
- The `add` method is the middle level that injects the dependencies into the dependency container using a class constant and a late-binding instantiating closure.
- The `get` method is the lowest level that sources the dependency.

**Conventions**

> - All class constants MUST be `public` (to decrease conflicts in definition for being a public API class).

Reviewer 6: Regardless of conflicts, they have to be public. Because, all injected dependencies are expected to be used in Factory classes. The class constant will be needed in Factory class.

```suggestion
- All class constants MUST be `public`.
```

- Each dependency MUST be defined using a class constant.
- Setting dependency using the `container::set()` MUST be paired with late-binding closure definition to decouple instantiation.

> - Dependencies MUST be wired through the target layer corresponding inherited method:

Reviewer 5: Dependencies that required to be instantiated every time we call it, e.g. Queries MUST be added via `Container::factory()` method. E.g. 
`protected function addUserPropelQuery(Container $container): Container
    {
        $container->set(static::PROPEL_QUERY_USER, $container->factory(function (): SpyUserQuery {
            return SpyUserQuery::create();
        }));

        return $container;
    }`

```php
public function provideCommunicationLayerDependencies(Container $container)
public function provideBusinessLayerDependencies(Container $container)
public function providePersistenceLayerDependencies(Container $container)
public function provideDependencies(Container $container)
public function provideBackendDependencies(Container $container)
public function provideServiceLayerDependencies(Container $container)
```
- Only three type of methods CAN be defined, either `provide`, `get`, or `add`.
```php
public function provide*Dependencies(Container $container)
```
- `Provide` methods MUST call their parent `provide` method.
- `Provide` methods MUST only call `add` methods.

```php
protected function add*(Container $container)
```
- `Add` methods MUST only introduce one dependency to the `Container` (a plugin-stack is considered one dependency in this respect).

```php
protected function get[Dependency](Container $container)
protected function get[PluginInterfaceName]Plugins(Container $container)
```

<details><summary markdown='span'>Additional Conventions for Project Development</summary>

> - Constants can have any access modifier, they are not limited to `public`. <br/>

Reviewer 6: Having protected and private access level does not bring any value. Is there any use-case where non-public access level will be useful?
```suggestion
```

Reviewer 7: yes, there are.
For example, we reuse a plugin stack from Calculate::quote in the Calculation::order, by unsetting/replacing some of the plugins by keys, and the keys are private/protected, since used only in this dependency provider.

Why projects should not be allowed to add some protected constant in this class?

Reviewer 6: Because in my mind, the class constants were supposed to be used in both dependency providers and factories. So, if you made them private/protected, they would be invisible to factories.

The use-case you shared is interesting one. I would group common plugins for quote and order inside of a method, and add new ones without the need of replacing, or unsetting.

In any-case this is a valid use case :). I will close my comment.

- Methods can have any access modifier. <br/>

> - `Provide` methods can call other methods than `add`. <br/>

Reviewer 6: Why does project development differ from the core development in this particular case?

```suggestion
- `Provide` methods can call other methods than `add`. <br/>
```

Author: it's the other way. can you tell me a reason why wouldn't it be allowed for project development?

Reviewer 6: You are right, there is no reason that I could think of to have such restriction.

I am just a little bit confused for whatever reason we have this restriction for the core development, it may also benefit the project development.

Author: @bnymn single responsibility. it's simpler to read the code. but this is not something that we can force on project without reason.

- `Add` methods can introduce more dependency to the `Container`.<br/>
- The `Container $container` argument can be introduced on need-to-have basis instead of being always mandatory.
</details>

<details><summary markdown='span'>Additional Conventions for Module Development</summary>
- [Facades](#facade-design-pattern) MUST be wrapped into a [Bridge](#bridge) to avoid coupling another module.<br/>
- [Plugins](#plugin) MUST NOT be wired on module level.<br/>

> - [Plugin](#plugin) defining methods MUST be tagged with `@api` tag as they are considered module API. <br/>

Reviewer 5: Dependency constant value must be the same as constant name

Reviewer 5: constants must have a prefix that describe its purpose. E.g. **FACADE**_STORE, **PLUGINS**_EXAMPLE, etc.

        });

        return $container;
    }

    /**
     * @api
     *
     * @param \Spryker\Zed\Kernel\Container $container
     *
     * @return array<\Spryker\Zed\ProductOfferExtension\Dependency\Plugin\ProductOfferPostUpdatePluginInterface>
     */
    protected function getProductOfferPostUpdatePlugins(Container $container): array
    {
        return [];
    }
}
```

## Entity

```
src
├── Generated
│   └── Shared
│       └── Transfer
│           └── [EntityName]EntityTransfer.php
│
├── Orm
│   └── Zed   
│       └── [DomainName]   
│           ├── Base
│           │   ├── [EntityName].php
│           │   └── [EntityName]Query.php 
│           ├── Map
│           │   └── [EntityName]TableMap.php
│           ├── [EntityName].php
│           └── [EntityName]Query.php
│
└── [Organisation]
    └── Zed
        └── [ModuleName]
            └── Persistence
                └── Propel
                    ├── Abstract[EntityName].php
                    └── Abstract[EntityName]Query.php
```

**Description**

Active record object which represents a row in a table. `Entities` have getter and setter methods to access the underlying data.

> Each `Entity` has a generated identical [Entity Transfer Object](#transfer-object) which can be used during interaction with other layers.

Reviewer 2: Something is not accurate here, we should never use Entity Transfer Objects anymore

Author: @demkos What is your opinion? I'm not aware of ETO restrictions 🤷

Reviewer 3: Entity transfers were limited to a single module API by this RFC: https://spryker.atlassian.net/wiki/spaces/RFC/pages/184254835/ACCEPTED+RFC+Scope+of+Entity+Transfers

**Example**

The lowest level class generated by Propel, containing the base Propel functionality. 
```php
namespace Orm\Zed\ConfigurableBundle\Persistence\Base;

/**
 * Base class that represents a row from the 'spy_configurable_bundle_template' table.
 */
abstract class SpyConfigurableBundleTemplate implements Propel\Runtime\ActiveRecord\ActiveRecordInterface {...}
```

The middle level class generated in a Spryker module, containing the SCOS module functionality.
```php
namespace Spryker\Zed\ConfigurableBundle\Persistence\Propel;

abstract class AbstractSpyConfigurableBundleTemplate extends Orm\Zed\ConfigurableBundle\Persistence\Base\SpyConfigurableBundleTemplate {...}
```

The highest level class generated on project, containing the project specific functionality.
```php
namespace Orm\Zed\ConfigurableBundle\Persistence;

class SpyConfigurableBundleTemplate extends Spryker\Zed\ConfigurableBundle\Persistence\Propel\AbstractSpyConfigurableBundleTemplate {...}
```

## Entity Manager

```
[Organization]
└── Zed
    └── [Module]
        └── Persistence
            ├── [Module]EntityManagerInterface.php
            └── [Module]EntityManager.php
```

**Description**

> Persists [Entities](#entity) by using their internal saving mechanism.

Reviewer 6: I don't think that we have to use Propel all the time with Entity Managers. Especially where performance is critical and there is a high load of data to be persisted.

```suggestion
Persists [Entities](#entity) either by using their internal saving mechanism, or by using raw SQL statements.
```

Reviewer 5: we should. Propel triggers P&S events and EntityManager is not responsible for it

Reviewer 6: When the data load is high, this pattern does not scale, and cause a lot of problems on production.

We should either suggest a new pattern for high data load, or accept that EntityManager can trigger events when raw SQL is used to persist entities.

Author: For core development, we are not ready to consider this freedom. however, i'm going to lift this restriction for project development. i believe it makes 100% sense 👍

The `Entity Manager` can be accessed from the same module's [business layer](#business-layer-responsibilities).

**Conventions**
- The entity manager class MUST define and implement an interface that holds the specification of each `public` method.

> - Entity manager `public` methods MUST have a functionality describing prefix, such as `create*()`, `delete*()`, `update*()`

Reviewer 6: ```suggestion
- Entity manager `public` methods MUST have a functionality describing prefix, such as `create*()`, `delete*()`, `update*()`.
```

- Creating, updating and deleting functions MUST be separated by concern even if they use overlapping internal methods.
- Entity manager methods MUST receive only [Transfer Objects](#transfer-object) as input parameters.

> - Entity manager methods MUST return `void` or the saved object(s) as [Transfer Object(s)](#transfer-object).

Author: @Vetaxon We don't have anything on this in our conventions. Could you please draft your idea?

<details><summary markdown='span'>Additional Conventions for Project Development</summary>

> - It is recommended to NOT define or maintain `Entity Manager` interfaces.<br/>

Author: @Vetaxon I wonder why you don't want to recommend to avoid usage?

     * - Returns fallback locale translation if provided single locale translation does not exist or translation key if nothing found.
     *
     * @api
     *
     * @param \Generated\Shared\Transfer\ConfigurableBundleTemplateFilterTransfer $configurableBundleTemplateFilterTransfer
     *
     * @return \Generated\Shared\Transfer\ConfigurableBundleTemplateResponseTransfer
     */
    public function getConfigurableBundleTemplate(
        ConfigurableBundleTemplateFilterTransfer $configurableBundleTemplateFilterTransfer
    ): ConfigurableBundleTemplateResponseTransfer;
}
```

## Facade design pattern

__TBD__

**Description**

Spryker defines the facade design pattern as the primary entry point for layers following the [standard facade design pattern](#https://en.wikipedia.org/wiki/Facade_pattern).

There are currently four components that use the facade design pattern:
- The [Facade](#facade) component that is the facade of the `Business layer`.
- The [Client](#client-facade) facade component that is the facade of the `Client application layer`.
- The [Service](#service-facade) facade component that is the facade of the `Service application layer`.

> - The [Query Container](#query-container-facade) facade component that is an overarching facade of the `Persistence layer`.

Author: i think we need to mention. to goal of this page is to give an overview what's happening, and it's important to be transparent.

  - return a [Transfer Object](#transfer-object) collection with the following naming pattern `[DomainEntity][Collection|CollectionResponse]Transfer`.
  - declare the `Create` and `Update` CUD methods per entity as follows: `[create|update][DomainEntity]Collection([DomainEntity]CollectionRequestTransfer): [DomainEntity]CollectionResponseTransfer`
  - declare the `Delete` CUD method as follows: `delete[DomainEntity]Collection([DomainEntity]CollectionDeleteCriteriaTransfer): [DomainEntity]CollectionResponseTransfer;`
- Multi-item-flow `Create` and `Update` methods MUST
  - support the transactional entity manipulation, and document it in facade specification.

__TBD__ (Note: Service is not serving data => no CRUD)

**Guidelines**
- The inherited `getFactory()` can be used to instantiate the underlying classes.
- The methods MAY throw exceptions.
- Single-item-flow methods SHOULD be avoided as they are NOT scalable.
- Multi-item-flow methods SHOULD operate in batches to be scalable.
- Multi-item-flow `Create` and `Update` methods SHOULD return a list of entities + error list.

<details><summary markdown='span'>Additional Conventions for Project Development</summary>

> - Definition of interface is NOT recommended. <br/>

Reviewer 10: According to the opinion of all tech leads on projects, it is better to use interfaces at all levels as in the core. We can say that this is not acceptable, but it is not recommended.

> - Definition of interface is NOT recommended. <br/>

Reviewer 9: I know that comments triggering debates are not welcomed in this review. 
But can you explain the use case and need of interfaces on ALL levels of project code? 

Why for example Pyz XxxModuleFacade should have an interface? Are they considering several implementations? 
@Vetaxon @profuel

> - Definition of interface is NOT recommended. <br/>

Reviewer 7: Locator works correctly without interface, to my surprise.
Do we want customers to write really shitty code without interfaces? My answer is NO.
Thus I liked and agree with the suggestion to remove the line that we do not recommend interface for entities that follow facade pattern.

@matweew I think we should encourage to write a good code, thus we do not force interfaces for internal models, but also do not discourage from interfaces on other levels.

> - Definition of interface is NOT recommended. <br/>

Reviewer 7: To help us understand your point, please answer the question - `why it's not recommended?`

> - Definition of interface is NOT recommended. <br/>

Reviewer 9: @profuel Code without interfaces != shitty code!

> Interfaces are powerful but not universally mandatory. Use them when they enhance code clarity, flexibility, and maintainability through polymorphism, loose coupling, and potential future expansion with multiple implementations.

So not every class needs interface(s):

>- **Simple Utility Classes**: For simple helper classes with no need for polymorphism or future expansion, an interface might be overkill.
>- **Single Responsibility Classes**: If a class has a very specific and well-defined purpose with no need for interchangeable behavior, an interface might not add much value.

I/we assume that 80-90% of classes on the project level fall under this criteria. That's why  `it's not recommended` to create extra code when there is no need in it.

The responsibility of a `Factory` is to instantiate classes of a layer and inject dependencies during instantiation.

**Conventions**
- `Factories` MUST NOT define/implement interface as practically they are never fully replaced.
- `Factories` MUST orchestrate the instantiation of objects in solitude (without reaching out to class-external logic).
- Factory methods MUST either instantiate one new object and named as `create[Class]()`, or wire the dependencies and named as `get[Class]()`.
- Factory methods MUST be `public`.
- Factory methods MUST use the constants defined in [dependency provider](#dependency-provider) when reaching out to module dependencies.

<details><summary markdown='span'>Additional Conventions for Project Development</summary>
- Factory methods can define/implement interfaces on demand.<br/>
- Factory methods can instantiate and write dependencies in one step, but this case MUST be named as `get[Class]()`.<br/>
- Factory methods can have any access modifier.
</details>

## Gateway Controller

```
[Organization]
└── Zed
    └── [Module]
        └── Communication
            └── Controller
                └── GatewayController.php
```

**Description**

> `Gateway Controller` is special, reserved `Controller` (see [Controllers](#controller)) in SCOS, that serves as an entry point in [Zed](#zed) for serving [Client](#client) requests (see [Zed Stub](#zed-stub) for more details).

Reviewer 5: I would also add that Gateway controller are used only from Zed or BackofficeGateway applications and are not accessible in Backoffice one

- [Persistence layer](#persistence-layer-responsibilities) `Models` are present in [Zed](#zed).

**Conventions**

- Each model class MUST define and implement an interface (`[ModelName]Interface.php`) that holds the specification of each `public` method.
- Dependencies MUST be injected via constructor.
  - `Models` MUST NOT instantiate any object but [Transfer Objects](#transfer-object).
- Dependencies MUST be referred/used/defined by interface.
- Dependencies MUST be from the same module, either `Models`, [Repository](#repository), [Entity Manager](#entity-manager), [Config](#module-configurations), or [Bridge wrapped facades](#bridge).
  - `Models` MUST NOT interact directly with other module's `Models` (eg: via inheritance, shared constants, instantiation, etc.).
- `Models` MUST be grouped under a folder by activity (eg: `Writers/ProductWriter.php`, `Readers/ProductReader.php`).
- `Models` MUST NOT be named using generic words, such as `Executor`, `Handler`, or `Worker`.
- `Model` methods MUST NOT be `private`.

<details><summary markdown='span'>Additional Conventions for Project Development</summary>
- Dependencies can be referred directly via class.

> - It is recommended to NOT define interface for `Models` without project specific reason.

Reviewer 10: ``` suggestion
- It is allowed NOT to define interface for `Models` without project specific reason.
```

Reviewer 9: I would leave "recommended" because it makes the codebase thin and when the need of an interface appears it is easy&quick to add it.

> - Facades can be directly used as dependency, `Bridge` wrapping is NOT recommended.

Reviewer 9: @profuel More code - more places for bugs, more code to maintain.

When `module configuration` is defined in another [application layer](#application-layers) than [Shared](#shared), they are dedicated and accessible to that single [application layer](#application-layers) only.

**Conventions**

- `Module configuration`
  - Getter methods (prefixed with `get`) MUST be defined to retrieve configuration values, instead of accessing values via constants directly (this enables a more flexible extension structure, and an easier to control access tracking).
    - `Protected` constants MAY also be defined to support the getter methods (example: to enable cross-module referencing via `@uses`).
- `Environment configuration`
  - constants MUST be always `public` and have specification about their purpose.
  - constants MUST contain the same UPPERCASE value as the key is + properly prefixed with MODULE_NAME (see example below).

<details><summary markdown='span'>Additional Conventions for Project Development</summary>
- `Module configuration` constants can be defined with any access modifier.

> - `Environment configuration` constants can be defined with any access modifier.

Author: Yet, i want to give freedom where possible. On project level, i don't find any core reason, why i would demand to follow `public only`.

**Example**

```php
class QuotePermissionChecker implements QuotePermissionCheckerInterface
{
use PermissionAwareTrait;

    /**
     * @uses \Spryker\Client\SharedCart\Plugin\ReadSharedCartPermissionPlugin::KEY
     */
    protected const PERMISSION_PLUGIN_KEY_READ_SHARED_CART = 'ReadSharedCartPermissionPlugin';

    public function checkQuoteReadPermission(QuoteTransfer $quoteTransfer): bool
    {
        ...
      return $this->can(
        static::PERMISSION_PLUGIN_KEY_READ_SHARED_CART,
        $quoteTransfer->getCustomer()->getCompanyUserTransfer()->getIdCompanyUser(),
        $quoteTransfer->getIdQuote(),
      );
    }
```

## Persistence Schema

```
[Organization]
└── Zed
    └── [Module]
        └── Persistence
            └── Propel
                └── Schema
                    └── spy_[domain_name].schema.xml
```

**Description**

The schema file defines the module's tables and columns (see [Propel Documentation - Schema XML](#https://propelorm.org/documentation/reference/schema.html)). Schema files are organized into business `domains`, each representing a module overarching group that encapsulates related domain entities.

> **Conventions**

Reviewer 5: identifierQuoting="true" must be added if field names can be used as a key words, e.g. `key`

Reviewer 7: I think this is the error that will be seen once you run a query.
Also, I assume it should be reported by Propel model builder.

Author: this is not Spryker convention but propel

> - Table name MUST follow the format `spy_[domain_entity_name]` (eg: `spy_customer_address`, `spy_customer_address_book`).

Reviewer 2: `spy_[domain_entity_name]` It's not true for the storage, search or relation (spy_cms_slot_to_cms_slot_template)

- Tables MUST have an integer ID primary key, following the format `id_[domain_entity_name]` (eg: `id_customer_address`, `id_customer_address_book`).
- Table definitions MUST include the `phpName` attribute.

> - Table foreign key MUST follow the format `fk_[remote_entity]` (eg: `fk_customer_address`, `fk_customer_address_book`).

Author: if you have 2 relations to the same column, the "old version" also doesn't help. :) but i will add some workaround for your case, good idea 👍

    </table>
</database>

```

Injecting columns from `Product` to a `Url` `domain` by defining `spy_url.schema.xml` schema file in `Product` module.
```xml
<?xml version="1.0"?>
<database xmlns="spryker:schema-01" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" name="zed" xsi:schemaLocation="spryker:schema-01 https://static.spryker.com/schema-01.xsd" namespace="Orm\Zed\Url\Persistence" package="src.Orm.Zed.Url.Persistence">

    <table name="spy_url">
        <column name="fk_resource_product_abstract" type="INTEGER"/>
        <foreign-key name="spy_url-fk_resource_product_abstract" foreignTable="spy_product_abstract" phpName="SpyProduct" onDelete="CASCADE">
            <reference foreign="id_product_abstract" local="fk_resource_product_abstract"/>
        </foreign-key>
    </table>
</database>
```

## Provider / Router

```
[Organization]
└── Yves
    └── [Module]  
        └── Plugin
            ├── Provider
            │   ├── [Name]ControllerProvider.php
            │   └── [Name]ServiceProvider.php
            └── [ConsumerModule]
                └── [RouterName]Plugin.php
```

**Description**

> `Providers` and `Routers` are used during the bootstrap of [Yves](#applications) application. There are three types of them:

Author: no. Zed is autowired. Glue has a different method that is not even described :) 🎉

- **Controller Provider** - Registers [Yves](#applications) [Controllers](#controller) of a module and binds them to a path.
- **Service Provider** - Represents a cross-functional [Yves](#applications) service (eg: a validator).
- **Router** - Resolves a path into a [Yves](#applications) [Controller and Action](#controller).

**Conventions**

> - All `providers` and `routers` MUST be registered in `\[Organization]\Yves\ShopApplication\YvesBootstrap.php`.

Reviewer 10: What about GlueBootstrap?

> - A `Controller Provider` MUST extend `\SprykerShop\Yves\ShopApplication\Plugin\Provider\AbstractYvesControllerProvider`.

Author: Almost. You are right, that the current content only covers Yves. But this is what we will need to fix on the long-term, to cover Zed & Glue as well.

> - A `Service Provider` MUST implement `\Silex\ServiceProviderInterface and may extend \Spryker\Yves\Kernel\AbstractPlugin`.

Reviewer 5: Please no. Sliex is deprecated and MUST NOT be used

Author: What do you mean by `please no` ?

Reviewer 7: Please do not add this line: 
> A `Service Provider` MUST implement `\Silex\ServiceProviderInterface`

- A `Router` MUST extend `\SprykerShop\Yves\ShopRouter\Plugin\Router\AbstractRouter`.
- A provider is classified as a [Plugin](#plugin) and so the plugin's conventions apply.

> ## Query Container facade

Author: no we can't, because it's not deprecated. we are still supporting it, and rolling out fixes for it. yet, we recommend it to be not implemented anymore, as this design has many flaws.

> ## Query Container facade

Author: but i've added an extra line, to use EM and Repository instead

```
[Organization]
└── Zed
    └── [Module]
        └── Persistence
            └── [Module]QueryContainer.php
```

**Description**

The `Query Container` is the internal API of [Persistence layer](#persistence-layer-responsibilities), and represents and entry point to database access.
The more advanced and modular [Entity Manager](#entity-manager) and [Repository](#repository) pattern was introduced to counter the problems of cross-module leaks of `Query Container` concept.

**Conventions**
- The `Query Container` follows the [facade design pattern](#facade-design-pattern) and its conventions.

> - New `QueryContainer` functionality MUST NOT be introduced but through the advanced [Entity Manager](#entity-manager) and [Repository](#repository) pattern.

Author: we recently were using the Query objects directly, rather than the Query Container facades. So it's definietly possible.

<details><summary markdown='span'>Additional Conventions for Project Development</summary>

> - `QueryContainers` can be used/developed further, but it is recommended to transition toward the [Entity Manager](#entity-manager) and [Repository](#repository) pattern.

Reviewer 10: ```suggestion
- `QueryContainers` can be used/developed further, but it is highly recommended to transition toward the [Entity Manager](#entity-manager) and [Repository](#repository) pattern.
```

      $query
        ->joinSpyUrl()
        ->joinSpyProductComment();
      ...
  }
```

## Repository

```
[Organization]
└── Zed
    └── [Module]
        └── Persistence
            ├── [Module]Repository.php      
            └── [Module]RepositoryInterface.php
```

**Descriptions**

Responsible for retrieving data from database by executing queries and returning the results as [Transfer Objects](#transfer-object) or native types.
The `Repository` can be accessed from the same module's [Communication](#communication-layer-responsibilities) and [Business Layers](#business-layer-responsibilities).

**Conventions**

- The `Repository` class MUST define and implement an interface (`[Model]RepositoryInterface.php`) that holds the specification of each `public` method.
- `Public` methods MUST receive and return [Transfer Objects](#transfer-object) only.
- The methods MUST return with a collection of items or a single item (eg: using and wrapping the results of `find()` or `findOne()`).
- `Repositories` MUST use [Query Objects](#query-object) to retrieve data from the database.
- `Repostiories` MUST NOT alter the data in the database, only retrieve it.

> 

Reviewer 10: ```suggestion
<details><summary markdown='span'>Additional Conventions for Project Development</summary>
- Methods can use also native php types as input arguments or result values.
</details>

```

Reviewer 8: I made a lot of refactoring and there is always a way not to do it

Reviewer 7: @vitaliiivanovspryker what about CTE?
`may` means that you are allowed to use queries to gain performance.

Author: but i agree with the proposal. there is no reason to forbid this for project development

> 

Reviewer 5: Can we also add Module convention that find methods are nullable and get are not?

Author: that's not part of this convention list. the general naming conventions etc. are located somewhere else

**Conventions**

- The bridge class MUST define and implement an interface (`[CurrentModule]To[FacadeOwnerModule][Service|Client|Facade]Interface.php`) that holds the specification of each `public` method (mind the missing bridge word!).
- `Bridges` MUST contain only the delegation logic to the friend facade method.
- The `Bridge` constructor MUST NOT contain parameters to avoid coupling to a specific class.
- `Bridge` classes and interfaces MUST be declared as strict as possible for input arguments and returns values, compared to their friend method with valid types (no mixed, no x|y).

**Guidelines**
- Bridge vs adapter: for simplification, we keep using bridge pattern even when adapting the earlier version of a SCOS facade. Adapters are used when the remote class' life-cycle is independent to SCOS or there is a huge technical difference between the adaptee and adaptor.
- During `Bridge` definitions, type definition mistakes in remote facades become more visible. In these cases, be aware of the cascading effect of changing/restricting an argument type in public API when you consider such changes.

> 

Reviewer 10: ```suggestion
<details><summary markdown='span'>Additional Conventions for Project Development</summary>
- We strongly discourage the usage of bridges in projects unless necessary due to business logic.
</details>
```

Author: You are in Module contribution section, nothing is applicable to project development.

│   └── [ConsumerModule]Extension
│       └── Dependency
│           └── Plugin
│               └── [PluginInterfaceName]PluginInterface.php
│
├── Glue
│   ├── [Module]
│   │   └── Plugin
│   │       └── [ConsumerModule]
│   │           └── [PluginName]Plugin.php    
│   └── [ConsumerModule]Extension
│       └── Dependency
│           └── Plugin
│               └── [PluginInterfaceName]PluginInterface.php
│
└── Client
    ├── [Module]        
    │   └── Plugin
    │       └── [ConsumerModule]
    │           └── [PluginName]Plugin.php    
    └── [ConsumerModule]Extension
        └── Dependency
            └── Plugin
                └── [PluginInterfaceName]PluginInterface.php

```

**Description**
- `Plugins` are classes which are used to realize Inversion-Of-Control - instead of a direct call to another module's [facade](#facade-design-pattern), a `plugin` can be provided as an optional and configurable class.
- `Plugins` are the only classes that can be instantiated across modules.
- See more details about the implementation in [Plugin Interface](#plugin-interface) component, and instantiation in [Dependency Provider](#dependency-provider).

**Conventions**

> - The `Plugin` MUST NOT contain business logic but delegates to the underlying [facade](#facade-design-pattern).

Reviewer 9: This rule adds a lot of overhead in the code and generates a lot of methods in the Facade for only one case (for one specific plugin). I vote for removing it or making it less strict.

For project Plugins the rule 100% must not be strict.

Reviewer 7: I like the idea, but from real life experience I can tell you it brings a lot of mess, especially for projects:
when plugin contains the logic, then tests are written against the plugin, and it gets wierd:
some plugins are tested , since they have business logic inside
some plugins are not tested, since the facade method they call is tested.
some plugins are tested with just a single facade method, and the facade method tested separately second time.

so my suggestion for projects is:
```suggestion
- The `Plugin` MAY contain business logic.
```

Author: Plugins are core concept. It is not recommended for projects to use Plugins at all. Therefore, nothing applies anyways.

Author: @matweew Although I'm fully on your side, we tried to remove this rule for many times and it triggers a huge debate. I think we can't remove it atm without aligning.

2. **Plugin executor**: The module that uses the [Plugin(s)](#plugin) in its [Dependency Provider](#dependency-provider) (example: `CompanyDependencyProvider::getCompanyPostCreatePlugins()` in `Company` module)
3. **Plugin providers**: The modules that implement a [Plugin](#plugin) (example: `CompanyBusinessUnitCreatePlugin` in `CompanyBusinessUnit` module)

**Conventions**
- `Plugin interfaces` MUST be defined in extension modules.
  - `Extension modules` MUST be suffixed with `Extension` and follow regular module architecture.
  - `Extension modules` MUST NOT contain anything else but `Plugin Interfaces` and [Shared application layer](#shared).
- `Plugin Interface` MUST have a class specification with plugin-stack details.
- Plugin interface methods MUST receive input items as collection for scalability.

**Guidelines**

- Operations on single items in plugin stack is forbidden, except for the following reasons:
  - it is strictly and inevitably a single-item flow.
  - the items go in FIFO order and there is no other way to use a collection instead.
- Plugin interface class specification SHOULD explain:

>   - how the [Plugins](#plugin)) will be used,

Reviewer 9: ```suggestion
  - how the [Plugins](#plugin) will be used,
```

  - what are the typical use-cases of a [Plugin](#plugin).

> 

Reviewer 10: ```suggestion
<details><summary markdown='span'>Additional Conventions for Project Development</summary>
- `Extension modules` are not mandatory for project development. 
</details>
```

Author: You are in Module contribution section. None of these components are related to project development.
